site/src/testHelpers/hooks.tsx

```diff
@@ -0,0 +1,193 @@
+  type RenderWithAuthOptions,
+  createTestQueryClient,
+} from "./renderHelpers";
+import {
+  type Location,
+  createMemoryRouter,
+  RouterProvider,
+  useLocation,
+} from "react-router-dom";
+
+export type RouterLocationSnapshot<TLocationState = unknown> = Readonly<{
+  search: URLSearchParams;
+  pathname: string;
+  state: Location<TLocationState>["state"];
+}>;
+
+export type GetLocationSnapshot<TLocationState = unknown> =
+  () => RouterLocationSnapshot<TLocationState>;
+
+export type RenderHookWithAuthResult<
+  TResult,
+  TProps,
+  TLocationState = unknown,
+> = Readonly<
+  Omit<RenderHookResult<TResult, TProps>, "rerender"> & {
+    queryClient: QueryClient;
+    rerender: (newProps: TProps) => Promise<void>;
+
+    /**
+     * Gives you back an immutable snapshot of the current location's values.
+     *
+     * As this is a snapshot, its values can quickly become inaccurate - as soon
+     * as a new re-render (even ones you didn't trigger yourself). Keep that in
+     * mind when making assertions.
+     */
+    getLocationSnapshot: GetLocationSnapshot<TLocationState>;
+  }
+>;
+
+export type RenderHookWithAuthConfig<TProps> = Readonly<{
+  routingOptions?: Omit<RenderWithAuthOptions, "children">;
+  renderOptions?: Omit<RenderHookOptions<TProps>, "wrapper">;
+}>;
+
+/**
+ * Gives you a custom version of renderHook that is aware of all our App
+ * providers (query, routing, etc.).
+ *
+ * Unfortunately, React Router does not make it easy to access the router after
+ * it's been set up, which can lead to some chicken-or-the-egg situations
+ * @see {@link https://github.com/coder/coder/pull/10362#discussion_r1380852725}
+ */
+export async function renderHookWithAuth<Result, Props>(
+  render: (initialProps: Props) => Result,
+  config: RenderHookWithAuthConfig<Props>,
+): Promise<RenderHookWithAuthResult<Result, Props>> {
+  /**
+   * Our setup here is evil, gross, and cursed because of how React Router
+   * itself is set up. We need to go through RouterProvider, so that we have a
+   * Context for calling all the React Router hooks, but that poses two
+   * problems:
+   * 1. <RouterProvider> does not accept children, so there is no easy way to
+   *    interface it with renderHook, which uses children as its only tool for
+   *    dependency injection
+   * 2. Even after you somehow jam a child value into the router, calling
+   *    renderHook's rerender method will not do anything. RouterProvider is
+   *    auto-memoized against re-renders, so because it thinks that its only
+   *    input (the router object) hasn't changed, it will stop the re-render,
+   *    and prevent any children from re-rendering (even if they would have new
+   *    values).
+   *
+   * Have to do a lot of work to side-step those issues, and make sure that
+   * we're not relying on internal React Router implementation details that
+   * could break at a moment's notice
+   */
+  let escapedLocation!: Location; // Definite assignment via !
```

Comment:
well that's a bit of typescript I've never seen before. I'm wary of comments that just say "this line of code means x" tho.

Reply:
It's less a "this line of code means x", and more a "here's a bigger warning sign for something that silences the compiler and is only a single, easy-to-miss character", but I get what you mean. Not the biggest fan either